src/Xamarin.Android.Build.Tasks/Tasks/JavaDependencyVerification.cs

```diff
@@ -0,0 +1,691 @@
+				log.LogMessage ("Found Java dependency '{0}:{1}' version '{2}' from AndroidLibrary '{3}'", art.GroupId, art.Id, art.Version, task.ItemSpec);
+				artifacts.Add (art);
+			}
+		}
+	}
+
+	public void AddPackageReferences (ITaskItem []? tasks)
+	{
+		foreach (var task in tasks.OrEmpty ()) {
+
+			// See if JavaArtifact/JavaVersion overrides were used
+			if (task.TryParseJavaArtifactAndJavaVersion ("PackageReference", log, out var explicit_artifact, out var attributes_specified)) {
+				artifacts.Add (explicit_artifact);
+				continue;
+			}
+
+			// If user tried to specify JavaArtifact or JavaVersion, but did it incorrectly, we do not perform any fallback
+			if (attributes_specified)
+				continue;
+
+			// Try parsing the NuGet metadata for Java version information instead
+			var artifact = finder?.GetJavaInformation (task.ItemSpec, task.GetMetadataOrDefault ("Version", string.Empty), log);
+
+			if (artifact != null) {
+				log.LogMessage ("Found Java dependency '{0}:{1}' version '{2}' from PackageReference '{3}'", artifact.GroupId, artifact.Id, artifact.Version, task.ItemSpec);
+				artifacts.Add (artifact);
+
+				continue;
+			}
+
+			log.LogMessage ("No Java artifact information found for PackageReference '{0}'", task.ItemSpec);
+		}
+	}
+
+	public void AddProjectReferences (ITaskItem []? tasks)
+	{
+		foreach (var task in tasks.OrEmpty ()) {
+			// See if JavaArtifact/JavaVersion overrides were used
+			if (task.TryParseJavaArtifactAndJavaVersion ("ProjectReference", log, out var explicit_artifact, out var attributes_specified)) {
+				artifacts.Add (explicit_artifact);
+				continue;
+			}
+
+			// If user tried to specify JavaArtifact or JavaVersion, but did it incorrectly, we do not perform any fallback
+			if (attributes_specified)
+				continue;
+
+			// There currently is no alternate way to figure this out. Perhaps in
+			// the future we could somehow parse the project to find it automatically?
+		}
+	}
+
+	public void AddIgnoredDependencies (ITaskItem []? tasks)
+	{
+		foreach (var task in tasks.OrEmpty ()) {
+			var id = task.ItemSpec;
+			var version = task.GetRequiredMetadata ("AndroidIgnoredJavaDependency", "Version", log);
+
+			if (version is null)
+				continue;
+
+			if (version != null && MavenExtensions.TryParseArtifactWithVersion (id, version, log, out var art)) {
+				log.LogMessage ("Ignoring Java dependency '{0}:{1}' version '{2}'", art.GroupId, art.Id, art.Version);
+				artifacts.Add (art);
+			}
+		}
+	}
+
+	bool TrySatisfyDependency (ResolvedDependency dependency)
+	{
+		if (!dependency.Version.HasValue ())
+			return artifacts.Any (a =>
+				a.GroupId == dependency.GroupId
+				&& a.Id == dependency.ArtifactId);
+
+		var dep_versions = MavenVersionRange.Parse (dependency.Version);
+
+		var satisfied = artifacts.Any (a =>
+			a.GroupId == dependency.GroupId
+			&& a.Id == dependency.ArtifactId
+			&& dep_versions.Any (r => r.ContainsVersion (MavenVersion.Parse (a.Version)))
+		);
```

Comment:
Should the `List<Artifact> artifacts` actually be a `Dictionary`? `Dictionary<(string groupId, string artifactId), Artifact>`?

So you can lookup existing entries without System.Linq?